Resolve FIXME about creating emptyURLs without Objective-C

LumioseSil opened the pull request:
#### fb9d435d895dc53d26bd894458a1cf8462940888
<pre>
Resolve FIXME about creating emptyURLs without Objective-C
<a href="https://bugs.webkit.org/show_bug.cgi?id=253248">https://bugs.webkit.org/show_bug.cgi?id=253248</a>

Reviewed by NOBODY (OOPS!).

We can create an empty url by passing an empty CFString to
CFURLCreateWithString.

* Source/WTF/wtf/URL.h:(emptyCFURL): Removed.

* Source/WTF/wtf/cf/URLCF.cpp: Inline emptyURL.

* Source/WTF/wtf/cocoa/URLCocoa.mm:(emptyURL): Removed.

* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp: Remove preprocessor and
  use CF functions to create an empty CFURL object.

* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm: Prevent passing a
  null array to CFURLCreateAbsoluteURLWithBytes.

Note: If there is some wacky setting or manual script somewhere that
forces Apple platforms to interpret this file as Objective-C, please
tell me where I can fix that now that it is no longer needed.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb9d435d895dc53d26bd894458a1cf8462940888

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1483 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1513 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1563 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2408 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1307 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1574 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1572 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1427 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1495 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1358 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1344 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2247 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1363 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1327 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1285 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1272 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1350 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1367 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2436 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1432 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1402 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1245 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/1522 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1318 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1334 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/320 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1447 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/1562 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/433 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
```diff
@@ -132,15 +132,14 @@ - (instancetype)initWithCoder:(NSCoder *)coder
-        if (!m_wrappedURL)
-            LOG_ERROR("Failed to decode NSURL due to invalid encoding of length %d. Substituting a blank URL", length);
+    if (hasBaseURL) {
+        RetainPtr<NSURL> baseURL = adoptNS((NSURL *)[coder decodeObjectOfClass:NSURL.class forKey:baseURLKey]);
```
This seems like an overrelease.

AZero13 replied:
How if I may ask?

whsieh replied:
`adoptNS` creates a `RetainPtr` in a way that only releases, without retaining. Thus, it's only used for +1 objects. From the method naming convention, it does not seem like `-decodeObjectOfClass:forKey:` returns a +1 object.

Thus, when the `RetainPtr` created using `adoptNS` falls out of scope, it will result in a `-release` with no paired `-retain` (or `+alloc`).

aproskuryakov commented on Source/WTF/wtf/cocoa/URLCocoa.mm:
```diff
@@ -52,9 +52,8 @@ - (BOOL)_web_looksLikeIPAddress;
 
 RetainPtr<CFURLRef> URL::emptyCFURL()
 {
-    // We use the toll-free bridge to create an empty value that is distinct from null that no CFURL function can create.
```
Was this comment incorrect? How exactly did you verify that this CFURLCreateWithString invocation is equivalent?

AZero13 replied:
Just to prove my point. I am going to do something. Watch this

AZero13 replied:
If the tests pass, then we know it is equivalent. Otherwise, the program will abort and all will crash down.

AZero13 commented on Source/WTF/wtf/cocoa/URLCocoa.mm:
```diff
@@ -52,9 +52,12 @@ - (BOOL)_web_looksLikeIPAddress;
-    return bridge_cast(adoptNS([[NSURL alloc] initWithString:@""]));
+    auto result = adoptCF(CFURLCreateWithString(kCFAllocatorDefault, CFSTR(""), nullptr));
+    auto result2 = adoptNS([[NSURL alloc] initWithString:@""]);
+    if (![result2 isEqual:bridge_cast(result.get())])
```
@aproskuryakov the fact these tests have passed just goes to show these are the same.

AZero13 replied:
I do not intend on keeping this abort code. This code was just to prove a point.

aproskuryakov replied:
I don't know if we had a test for this behavior, or why it was needed. Not failing any tests is a useful data point, but no proof. 

The standard of prof is fairly high here, as the benefit of changing this code at all is unclear. And even if we were certain that the change was correct, we'd want to understand that it's beneficial in some way.

AZero13 replied:
Well, it would be nearly impossible for the tests to pass if an abort() were to have happened at any time, now would it?

LumioseSil committed:
Resolve FIXME about creating emptyURLs without Objective-C
https://bugs.webkit.org/show_bug.cgi?id=253248

Reviewed by NOBODY (OOPS!).

We can create an empty url by passing an empty CFString to
CFURLCreateWithString.

* Source/WTF/wtf/URL.h:(emptyCFURL): Removed.

* Source/WTF/wtf/cf/URLCF.cpp: Inline emptyURL.

* Source/WTF/wtf/cocoa/URLCocoa.mm:(emptyURL): Removed.

* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp: Remove preprocessor and
  use CF functions to create an empty CFURL object.

* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm: Prevent passing a
  null array to CFURLCreateAbsoluteURLWithBytes.

Note: If there is some wacky setting or manual script somewhere that
forces Apple platforms to interpret this file as Objective-C, please
tell me where I can fix that now that it is no longer needed.
